### PR TITLE
Product property building refactoring

### PR DESCRIPTION
**Pull request by 0nko**
This PR moves all the product detail property-building logic out of the `ProductDetailFragment` into a separate `ProductDetailCardBuilder` class. All the property types are abstracted away from the Android views and the hierarchy is now generated inside the ViewModel. The view hierarchy is constructed using nested `RecyclerViews`, which allows for a lot of flexibility. The same approach was used in Stats refresh in WPAndroid and it worked fairly well.

This whole thing has a couple benefits:

- The UI is modelled using simple data classes, which means the Product detail view hierarchy can be unit tested. The previous implementation had a lot of decision logic based on the product state from the ViewModel, which was breaking the rule that views should only be responsible for displaying data.
- The `ProductDetailFragment` was becoming too large, reaching almost 1000 LOC. Now, every property type has its own model and ViewHolder class, which are easier to navigate.
- New property types can be added easily in a modular way.
- All property models can now be reused on screens other than Product details, such as Product variations. This was my main motivation for the refactoring. All we need is to have a RecyclerView and a CardBuilder.

I know this PR is pretty massive but it wasn't really possible to split it up. I'd like to merge it to `develop` because the chance for merge conflict is very high. With that in mind, I'd like to ask everyone to please smoke test the Product details screen thoroughly, so that no new bugs are introduced. I've tried be careful to follow the old code, but you know how it is... :)

Any suggestions welcome!

### CODE REVIEW

**Comment by nbradbury** [comment]
Looks like the description is no longer rendering HTML. This description has bolded words but they're not showing up:

pr

In `develop` the bolding is working:

Screenshot_1588768975

**Comment by 0nko** [comment]
> Looks like the description is no longer rendering HTML

@nbradbury Fixed.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/WCProductPropertyView.kt [hunk 1]
```diff
@@ -62,9 +62,14 @@ class WCProductPropertyView @JvmOverloads constructor(
         onClickListener?.let {
             propertyGroupImg?.visibility = View.VISIBLE
             view?.setOnClickListener(onClickListener)
+            this.isClickable = true
         }
     }
 
+    fun removeClickListener() {
+        this.isClickable = false
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/WCProductPropertyView.kt:70]: Should this also nullify the assigned click listener, or does that not matter?

**Comment by nbradbury** [comment]
@0nko There's some very noticeable redrawing and refreshing when returning to product detail from product settings. The refreshing also happens after updating the product.

redraw

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/WCProductPropertyCardView.kt [hunk 2]
```diff
@@ -30,5 +35,24 @@ class WCProductPropertyCardView @JvmOverloads constructor(
             captionTextView.text = caption
             divider.show()
         }
+
+        propertiesRecyclerView.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/WCProductPropertyCardView.kt:39]: We can probably skip this if the layout manager is already assigned.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/WCProductPropertyCardView.kt [hunk 3]
```diff
@@ -30,5 +35,24 @@ class WCProductPropertyCardView @JvmOverloads constructor(
...
+
+        loadData(properties)
+    }
+
+    private fun loadData(data: List<ProductProperty>) {
+        val recyclerView: RecyclerView = view.findViewById(R.id.propertiesRecyclerView)
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/WCProductPropertyCardView.kt:45]: I think we can skip `findViewById` here and just refer to `propertiesRecyclerView`.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/adapters/ProductPropertiesDiffCallback.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import com.woocommerce.android.ui.products.models.ProductProperty.Type.READ_MORE
+
+class ProductPropertiesDiffCallback(
+    private val oldList: List<ProductProperty>,
+    private val newList: List<ProductProperty>
+) : Callback() {
+    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
+        val newItem = newList[newItemPosition]
+        val oldItem = oldList[oldItemPosition]
+        return if (oldItem.type == newItem.type) {
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/adapters/ProductPropertiesDiffCallback.kt:21]: I'm wondering if this `when` clause is necessary. Can't we just check `oldItem == newItem`?

**Comment by nbradbury** [comment]
@0nko This all looks very slick! I left a few nitpicks in the comments, but I do think the refresh/redraw issue will need to be addressed before merging.

I'm also not sure why tests are failing, since as we said on Slack they're working for us.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/adapters/ProductPropertiesDiffCallback.kt (hunk 4 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/adapters/ProductPropertiesDiffCallback.kt]: It's not. Good call.

**Comment by 0nko** [comment]
@nbradbury Thank you for the review! I've addressed your code comments and the update flashing should be fixed as well now. Ready for another pass.

Regarding the failing tests, I'm not really sure what to do. Maybe @oguzkocer could give us some hints? (TLDR; locally all unit tests are passing, but some are failing on CircleCI)

**Comment by oguzkocer** [comment]
@0nko Can you try running `./gradlew test` and verify that all tests pass for you? I see the exact same failures that CI reports even though the tests pass when I run it from Android Studio.

**Comment by 0nko** [comment]
Thanks @oguzkocer. After hours of investigations it turned out it was caused by a feature flag being enabled in Debug mode only, which made some tests fail in the Release mode 🤦.

@nbradbury All the comments should be addressed now and the unit tests are passing. Ready for another round!

**Pull request merged**